lib/silkpre/precompile.cpp

```diff
@@ -500,3 +506,53 @@ const SilkpreContract kSilkpreContracts[SILKPRE_NUMBER_OF_ISTANBUL_CONTRACTS] =
     {silkpre_bn_mul_gas, silkpre_bn_mul_run},     {silkpre_snarkv_gas, silkpre_snarkv_run},
     {silkpre_blake2_f_gas, silkpre_blake2_f_run},
 };
+
+SilkpreResult ethprecompiled_ecrecover(const uint8_t* input, size_t input_size, uint8_t* output, size_t output_size) {
+    assert(output_size == 32);
+    auto r = silkpre_ecrec_run(input, input_size);
+    std::copy_n(r.data, output_size, output);
+    return {0, r.size};
+}
+
+SilkpreResult ethprecompiled_sha256(const uint8_t* input, size_t input_size, uint8_t* output, size_t output_size) {
+    assert(output_size == 32);
+    silkpre_sha256(output, input, input_size, /*use_cpu_extensions=*/true);
+    return {0, 32};
+}
+
+SilkpreResult ethprecompiled_ripemd160(const uint8_t* input, size_t input_size, uint8_t* output, size_t output_size) {
+    assert(output_size == 32);
+    std::memset(output, 0, 12);
+    silkpre_rmd160(&output[12], input, input_size);
+    return {0, 32};
+}
+
+SilkpreResult ethprecompiled_expmod(const uint8_t* input, size_t input_size, uint8_t* output, size_t output_size) {
+    const auto r = silkpre_expmod_run(input, input_size);
+    assert(output_size == r.size);
```

Comment:
We are debating how to handle the case where user provides output buffer not big enough to handle the output produced by the library. This situation is **programmer error** caused by incorrectly computing the max output size and must be fixed in code. Reasonable handling is to abort/panic the process.

1. The `output_size` parameter may be removed and library should assume the buffer is big enough. Related error can only be detected by crashes or ASan tools. This is also not compatible with Rust `from_raw_parts()`.
2. The library aborts execution with some log.
3. The library reports the special error code which must be handled differently than EVM exceptions. The client does not have much options than abort in this case.